### PR TITLE
Add prototype of topline/executive summary

### PR DESCRIPTION
**Pull request by acmiyaguchi**
This is a prototype of what the finalized topline summary will look like (but rewritten in scala) for [bug 1309574](https://bugzilla.mozilla.org/show_bug.cgi?id=1309574). This could be a good example for other people trying to figure out how to use sparksql.

r? @mreid-moz @vitillo

### CODE REVIEW

**Comment by vitillo** [comment]
@mreid-moz r?

File: etl/topline_summary.kp/knowledge.md [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+
+```python
+import arrow
+
+def date_range(start_date, period, last_period=False):
+    """Returns the begin_date, end_date for this report.
+    
+    start_date:  datestring in the format YYYYMMDD
+    period:      type of period, either 'months' or 'weeks'
+    last_period: determines if this will this or last month, defaults to this month
```
Comment (mreid-moz) [inline etl/topline_summary.kp/knowledge.md]: I think there's a word missing here.

File: etl/topline_summary.kp/knowledge.md [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+search_schema = StructType([
+    StructField("google", IntegerType(), False),
+    StructField("bing", IntegerType(), False),
+    StructField("yahoo", IntegerType(), False),
+    StructField("other", IntegerType(), False)])
+
+search_pat = ['[Gg]oogle', '[Bb]ing', '[Yy]ahoo', '.']
+search_pat_comp = [re.compile(pat) for pat in search_pat]
+
+def get_search_counts(row):
```
Comment (mreid-moz) [inline etl/topline_summary.kp/knowledge.md]: These helper functions would be good candidates to move over to `standards.py` in [python_moztelemetry](https://github.com/mozilla/python_moztelemetry).

File: etl/topline_summary.kp/knowledge.md [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+    easy_aggregate_df = (
+        report_df
+            .groupby(groups)
+            .agg(
+                F.sum('hours').alias('hours'), 
+                F.sum('search_counts.google').alias('google'),
+                F.sum('search_counts.bing').alias('bing'),
+                F.sum('search_counts.yahoo').alias('yahoo'),
+                F.sum('search_counts.other').alias('other')) 
+            .join(crash_agg_df, ['country', 'channel', 'os', 'date'])
```
Comment (mreid-moz) [inline etl/topline_summary.kp/knowledge.md]: `crash_agg_df` doesn't appear to have a `date` column, does that impact this join (or the `groupby` that uses that column as well)?

File: etl/topline_summary.kp/knowledge.md [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+    
+    return easy_aggregate_df
+```
+
+```python
+def get_client_values(report_df, groups, report_date):
+    """ Return the number of new clients, default clients, and active users. """
+    
+    # Used to determine new clients in the report period
+    report_timestamp = arrow.get(report_date, 'YYYYMMDD').timestamp
```
Comment (mreid-moz) [inline etl/topline_summary.kp/knowledge.md]: The `YYYYMMDD` format is repeated in a couple of places, it should probably be factored out to a single place.

File: etl/topline_summary.kp/knowledge.md [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+# All report data is grouped by and then aggregated over the following columns
+groups = ['country', 'channel', 'os', F.lit(report_date).alias('date')]
+
+# Get the aggregated dataframes
+easy_aggregates_df = get_easy_aggregates(report_df, crash_df, groups, report_date)
+client_values_df = get_client_values(report_df, groups, report_date)
+
+# Lets join these dataframes together
+final_report_df = easy_aggregates_df.join(client_values_df, ['country', 'channel', 'os', 'date'])
+final_report_df.show()
```
Comment (mreid-moz) [inline etl/topline_summary.kp/knowledge.md]: Should we save the resulting data somewhere?

File: etl/topline_summary.kp/knowledge.md [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+created_at: 2016-11-14 00:00:00
+updated_at: 2016-11-30 16:34:01.126578
+tldr: 'Port of the topline/executive summary using the main summary dataset (not for
+  production usage).
+
+  '
+---
+# Topline Summary using Main Summary Dataset
+
+[Bug 1309574](https://bugzilla.mozilla.org/show_bug.cgi?id=1309574) involves porting the executive summary to use the main summary dataset. The original script can be found [mozilla-services/data-pipeline/run_executive_report.py](https://github.com/mozilla-services/data-pipeline/blob/e5c29541794325388336a210746029dce998b9e5/reports/executive_summary/run_executive_report.py).
```
Comment (mreid-moz) [inline etl/topline_summary.kp/knowledge.md]: Please also add a comment indicating, for each field in the expected output, whether that field is implemented here or not.

**Comment by vitillo** [comment]
@acmiyaguchi Are you still working on this PR?

**Comment by acmiyaguchi** [comment]
I'm not actively working on this at the moment, I would like to get around to it about the time that the scala version is done.

**Pull request closed without merging**